design/feature-flags.md

```diff
@@ -0,0 +1,92 @@
+existing projects.
+
+When we release a new major version of the AWS CDK, we will flip this behavior
+or completely remove the legacy behavior.
+
+In order for new projects to pick up this new behavior automatically, we will
+modify `cdk init` to inject the set of "future flags" into the generated
+`cdk.json` file. This means that the new project will have the latest behavior,
+but projects that were created prior to the introduction of this feature will
+have the same legacy behavior based on the set of capabilities that were
+available at the time of the project's creation. This list will be cleaned up
+every time we release a major version of course.
+
+Using fine-grained flags will allow users of old projects to pick up specific
+new "future" behaviors by manually adding the specific keys to their `cdk.json`
+file, without risking breakage in other unexpected areas.
+
+## Alternative Considered
+ 
+We considered an alternative of "bundling" new capabilities under a single flag
+that specifies the CDK version which created the project, but this means that
+users won't have the ability to pick and choose which future capabilities they
+want to enable in case they need them but don't want to take the risk of
+unexpected changes.
+ 
+The downside of the fine-grained approach is that it could result in a "blowing
+up" new `cdk.json` files in case there will be many new breaking capabilities
+between major releases. But this is hypothetical and even if this list ends up
+with 20 features before we release the next major version, I still think the
+benefits outweigh the risks of the alternative approach.
+
+## Design
+
+Context keys for feature flags will be listed in `cx-api/lib/features.ts` and
+will take the form: `<module>:<feature>`. 
+
+For example:
+
+- `@aws-cdk/core:enableStackNameDuplicates`
+- `@aws-cdk/aws-cloudformation:doNotCapitalizeCustomResourcePropertyNames`.
+
+Using the module name will allow easy tracing of the code that consumes this
+flag.
+
+The configuration for which feature flags should be enabled for new projects
+will be under `cx-api/lib/future.ts` and will be encoded as a simple context
+hash that will be injected by `cdk init` to all `cdk.json` files generated for
+new projects.
+
+We will mandate that when a feature or bug fix is introduced under a feature
+flag, the CHANGELOG will include:
+
+- The suffix `(under feature flag)` in the title.
+- A `BREAKING CHANGES` paragraph will be added which describes the *new*
+  behavior but disclaims that it will only apply to new projects created through
+  `cdk init`. It will also indicate the context key this flag uses for users who
+  wish to enable it manually in their project.
+
+## Future Development
+
+As a general rule, using a feature flag should be last resort in the case where
+it is impossible to implement backwards compatibility. A feature flag is likely
+to get less usage and therefore mature slower, so it's important to make sure we
+don't abuse this pattern.
```

Comment:
It would be good to get deeper into the criteria for when it's appropriate to use a feature flag. Might also be a good idea to strengthen the requirements for using one (additional reviewers, mini RFC, etc). Any mechanism that would mitigate the risk you've called out of abusing the pattern.

Reply:
Isn't "impossible to implement backwards compatibility" a sufficient criteria? Are you concerned that not all backwards compatibility avenues will be considered?

Reply:
fair point, but yes my thinking was that this would be a quick escape from due diligence. Impossible is clear, but there are likely going to be scenarios where it's counter-intuitive to implement a backwards compatible change and preferred to introduce through a feature flag. That line is blurrier than the "impossible" criteria.

Reply:
I honestly don't know how to define this better. It's case-by-case. The best I can propose is that we put a high bar on when a feature flag is approved (maybe we can say that at least two core team members are required to approve this.)